cdap-master/src/main/java/co/cask/cdap/data/tools/MDSUpgrader.java

```diff
@@ -0,0 +1,347 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.data.tools;
+
+import co.cask.cdap.api.common.Bytes;
+import co.cask.cdap.api.data.stream.StreamSpecification;
+import co.cask.cdap.api.dataset.DatasetDefinition;
+import co.cask.cdap.api.dataset.DatasetProperties;
+import co.cask.cdap.api.dataset.table.Row;
+import co.cask.cdap.api.dataset.table.Scanner;
+import co.cask.cdap.api.dataset.table.Table;
+import co.cask.cdap.app.ApplicationSpecification;
+import co.cask.cdap.app.runtime.ProgramController;
+import co.cask.cdap.app.store.Store;
+import co.cask.cdap.common.conf.CConfiguration;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.data2.datafabric.dataset.DatasetsUtil;
+import co.cask.cdap.data2.dataset2.DatasetFramework;
+import co.cask.cdap.data2.dataset2.lib.table.MDSKey;
+import co.cask.cdap.data2.dataset2.tx.DatasetContext;
+import co.cask.cdap.data2.dataset2.tx.Transactional;
+import co.cask.cdap.internal.app.store.AppMetadataStore;
+import co.cask.cdap.internal.app.store.ApplicationMeta;
+import co.cask.cdap.internal.app.store.DefaultStore;
+import co.cask.cdap.internal.app.store.ProgramArgs;
+import co.cask.cdap.proto.Id;
+import co.cask.cdap.proto.ProgramRunStatus;
+import co.cask.cdap.proto.ProgramType;
+import co.cask.cdap.proto.RunRecord;
+import co.cask.tephra.TransactionExecutor;
+import co.cask.tephra.TransactionExecutorFactory;
+import com.google.common.base.Joiner;
+import com.google.common.base.Supplier;
+import com.google.common.base.Throwables;
+import com.google.common.collect.Sets;
+import com.google.inject.Inject;
+import com.google.inject.name.Named;
+import org.apache.twill.filesystem.Location;
+import org.apache.twill.filesystem.LocationFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Set;
+
+/**
+ * Upgraded the Meta Data for applications
+ */
+public class MDSUpgrader extends AbstractUpgrader {
+
+  private static final Logger LOG = LoggerFactory.getLogger(MDSUpgrader.class);
+  private final Transactional<DatasetContext<Table>, Table> appMDS;
+  private final CConfiguration cConf;
+  private final Store store;
+  private final Set<String> appStreams;
+
+  @Inject
+  private MDSUpgrader(LocationFactory locationFactory, TransactionExecutorFactory executorFactory,
+                      @Named("dsFramework") final DatasetFramework dsFramework, CConfiguration cConf,
+                      @Named("defaultStore") final Store store) {
+    super(locationFactory);
+    this.cConf = cConf;
+    this.store = store;
+    this.appMDS = Transactional.of(executorFactory, new Supplier<DatasetContext<Table>>() {
+      @Override
+      public DatasetContext<Table> get() {
```

Comment:
Just the old table getting part or the whole MDS upgrade logic ? Last time when we discussed this we planned that only for log and for MDS since it just makes some constants public we decided its fine to keep it out.